packages/mobile/src/navigator/QRNavigator.tsx

```diff
@@ -0,0 +1,136 @@
+import {
+  createMaterialTopTabNavigator,
+  MaterialTopTabBarProps,
+} from '@react-navigation/material-top-tabs'
+import { useIsFocused } from '@react-navigation/native'
+import React, { useEffect, useMemo, useRef, useState } from 'react'
+import { useAsync } from 'react-async-hook'
+import { useTranslation } from 'react-i18next'
+import { Dimensions, Platform, StatusBar, StyleSheet } from 'react-native'
+import { check, PERMISSIONS, RESULTS } from 'react-native-permissions'
+import Animated, { call, greaterThan, onChange } from 'react-native-reanimated'
+import { ScrollPager } from 'react-native-tab-view'
+import { Namespaces } from 'src/i18n'
+import { Screens } from 'src/navigator/Screens'
+import QRCode from 'src/qrcode/QRCode'
+import QRScanner from 'src/qrcode/QRScanner'
+import QRTabBar from 'src/qrcode/QRTabBar'
+import { SVG } from 'src/send/actions'
+import { ExtractProps } from 'src/utils/typescript'
+
+const Tab = createMaterialTopTabNavigator()
+
+const width = Dimensions.get('window').width
+const initialLayout = { width }
+
+type AnimatedScannerSceneProps = ExtractProps<typeof QRScanner> & {
+  position: Animated.Value<number>
+}
+
+// Component doing our custom transition for the QR scanner
+function AnimatedScannerScene({ position, ...props }: AnimatedScannerSceneProps) {
+  const isFocused = useIsFocused()
+  const [wasFocused, setWasFocused] = useState(isFocused)
+  const [isPartiallyVisible, setIsPartiallyVisible] = useState(false)
+  const cameraPermission = useAsync(check, [
+    Platform.select({ ios: PERMISSIONS.IOS.CAMERA, default: PERMISSIONS.ANDROID.CAMERA }),
+  ])
+  // DENIED means the permission has not been requested / is denied but requestable
+  const hasAskedCameraPermission =
+    cameraPermission.result !== undefined && cameraPermission.result !== RESULTS.DENIED
+
+  useEffect(() => {
+    if (isFocused && !wasFocused) {
+      setWasFocused(true)
+    }
+  }, [isFocused])
+
+  Animated.useCode(
+    () =>
+      onChange(
+        greaterThan(position, 0),
+        call([position], ([value]) => {
+          setIsPartiallyVisible(value > 0)
+        })
+      ),
+    [position]
+  )
+
+  const animatedStyle = useMemo(() => {
```

Comment:
one thing ive been thinking about is defining big hooks like this outside the component and then just calling them. 

at best its more performant (no proof) at worst it keeps the voice in my head that functions should be 10 lines or less feeling less ignored.

Reply:
yes, kinda torn with this too, I'd say it depends, sometimes it's good to have all the context in a "big" function.